### PR TITLE
Set the correct level in the userlevel

### PR DESCRIPTION
**Pull request by balderdash**
Also grab the best progress when rendering. Not sure how I thought
this was working without the js changes here.

There's some logic duplicated across ruby and javascript for finding the level with the most progress, avoiding that would take a pretty large refactor :/

### CODE REVIEW

**Comment by Hamms** [comment]
I'm slightly confused. Why does the client need to know about the existence of multiple levels? Shouldn't it just handle whatever level the server gives it, and then the server handle processing the result with knowledge of multiple levels?

**Comment by balderdash** [comment]
For milestone posts, that could work for plain level swaps (just determine the level in the activities controller the same way we do in the scriptlevels controller), but it wouldn't be sufficient for CYOA.
@cpirich is also making changes that require an explicit level ID in milestone posts, so it's easier to just use that.

For rendering progress, it could go either way. E.g. I could have just sent a single status for each scriptlevel id instead of each level id. That was just a more complicated change though (I'd also have to reconsider the way we handle multi-page assessments), so I went with adding multiple levels to the scriptlevel summary and having the client figure it out instead.

File: code-studio/src/js/progress.js [hunk 1]
```diff
@@ -153,6 +154,31 @@ progress.renderCourseProgress = function (scriptData, currentLevelId) {
...
+// Return the level with the highest progress, or the first level if none have
+// been attempted
+progress.bestResultLevelId = function (levelIds, progress) {
+  // The usual case
+  if (levelIds.length === 1) {
+    return levelIds[0];
+  }
+
+  // Return the level with the highest result
+  var attemptedIds = levelIds.filter(function (id) { progress[id] !== 0; });
```
Comment (Hamms) [inline code-studio/src/js/progress.js]: this either needs a `return` OR (preferred) to just be an arrow function: `levelIds.filter(id => progress[id] !== 0);`

File: code-studio/src/js/progress.js [hunk 2]
```diff
@@ -153,6 +154,31 @@ progress.renderCourseProgress = function (scriptData, currentLevelId) {
...
+  if (attemptedIds.length === 0) {
+    // None of them have been attempted, just return the first
+    return levelIds[0];
+  }
+  var bestId = levelIds[0];
+  var bestResult = progress[bestId];
+  levelIds.each(function (id) {
```
Comment (Hamms) [inline code-studio/src/js/progress.js]: s/each/forEach

File: code-studio/test/js/progressTest.js [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
 var assert = require('assert');
 var progress = require('../../src/js/progress');
+var serverProgress;
```
Comment (Hamms) [inline code-studio/test/js/progressTest.js]: is there a reason this is here and not inside the relevant `describe`?

File: dashboard/app/models/user.rb [hunk 4]
```diff
@@ -835,8 +836,8 @@ def User.track_proficiency(user_id, script_id, level_id)
 
   # returns whether a new level has been completed and asynchronously enqueues an operation
   # to update the level progress.
-  def track_level_progress_async(script_level:, new_result:, submitted:, level_source_id:)
-    level_id = script_level.level_id
+  def track_level_progress_async(script_level:, level:, new_result:, submitted:, level_source_id:)
```
Comment (Hamms) [inline dashboard/app/models/user.rb:838]: are we fully confident we've updated all places where this is being used?

File: dashboard/app/serializers/user_serializer.rb [hunk 5]
```diff
@@ -38,6 +38,7 @@
 #  confirmation_sent_at       :datetime
 #  unconfirmed_email          :string(255)
 #  prize_teacher_id           :integer
+#  hint_access                :boolean
```
Comment (Hamms) [inline dashboard/app/serializers/user_serializer.rb]: can I assume this is an artifact of a migration from another PR?

File: dashboard/test/controllers/activities_controller_test.rb [hunk 6]
```diff
@@ -168,6 +168,34 @@ def _test_logged_in_milestone(async_activity_writes:)
...
+    params[:script_level_id] = script_level.id
+    params[:level_id] = level1.id
+    params[:result] = 'true'
+
+    post :milestone, params
+
+    assert_equal level1, UserLevel.last.level
+  end
+
+  test "successful milestone updates second userlevel when scriptlevel has multiple levels" do
```
Comment (Hamms) [inline dashboard/test/controllers/activities_controller_test.rb]: something like "updates specified userlevel" might be a more accurate title

**Comment by Hamms** [comment]
This looks good although it is slightly scary. I'm confused as to why all the tests passed even though I would expect JS errors. Do we understand why that might be?

**Comment by balderdash** [comment]
Hmm, pretty sure it's just a lack of test coverage. I'll add some JS tests before merging.

File: code-studio/src/js/progress.js (hunk 1 above)
Comment (balderdash) [inline code-studio/src/js/progress.js]: I need to get better at switching between ruby and javascript. Done.

File: code-studio/src/js/progress.js (hunk 2 above)
Comment (balderdash) [inline code-studio/src/js/progress.js]: Done.

File: code-studio/test/js/progressTest.js (hunk 3 above)
Comment (balderdash) [inline code-studio/test/js/progressTest.js]: Nope, moved.

File: dashboard/app/models/user.rb (hunk 4 above)
Comment (balderdash) [inline dashboard/app/models/user.rb:838]: Yeah, `git grep` says it only has one real call, and one more in a test.

File: dashboard/app/serializers/user_serializer.rb (hunk 5 above)
Comment (balderdash) [inline dashboard/app/serializers/user_serializer.rb]: Yep, removed.

File: dashboard/test/controllers/activities_controller_test.rb (hunk 6 above)
Comment (balderdash) [inline dashboard/test/controllers/activities_controller_test.rb]: Done.

**Comment by balderdash** [comment]
Alright, the tests were passing because:

a) The missing `return` in the filter call meant `attemptedIds` was always empty, so `bestResultsLevelId` returned early.

b) `assert` is not `assert.strictEqual`. It was just checking that the return value was truthy, treating the expected value as the failure message.

If I fix those two and revert the s/each/forEach/ change, it fails as expected. The world makes sense again.

The other tests in that file also haven't been testing what they were supposed to be testing this whole time. Fixed them as well.

**Comment by Hamms** [comment]
Hah, nice catch; that's frighteningly subtle.

LGTM!

**Pull request merged**